test/Sema/superfluously-public-imports.swift

```diff
@@ -0,0 +1,157 @@
+public import UnusedImport // expected-warning {{public import of 'UnusedImport' was not used in public declarations or inlinable code}} {{1-7=internal}}
+// expected-note @-1 {{imported 'public' here}}
+
+//--- Client.swift
+public import DepUsedFromInlinableCode
+public import DepUsedInSignature
+public import Exporter
+public import ConformanceBaseTypes
+public import ConformanceDefinition
+public import AliasesBase
+public import Aliases
+
+public import UnusedImport // expected-warning {{public import of 'UnusedImport' was not used in public declarations or inlinable code}} {{1-8=}}
+public import UnusedImport // expected-warning {{public import of 'UnusedImport' was not used in public declarations or inlinable code}} {{1-8=}}
+package import UnusedImport // expected-warning {{package import of 'UnusedImport' was not used in package declarations}} {{1-9=}}
+
+package import UnusedPackageImport // expected-warning {{package import of 'UnusedPackageImport' was not used in package declarations}} {{1-9=}}
+public import ImportNotUseFromAPI // expected-warning {{public import of 'ImportNotUseFromAPI' was not used in public declarations or inlinable code}} {{1-8=}}
+public import ImportUsedInPackage // expected-warning {{public import of 'ImportUsedInPackage' was not used in public declarations or inlinable code}} {{1-7=package}}
+
+public func useInSignature(_ a: TypeUsedInSignature) {} // expected-remark {{struct 'TypeUsedInSignature' is imported via 'DepUsedInSignature'}}
+public func exportedTypeUseInSignature(_ a: ExportedType) {} // expected-remark {{struct 'ExportedType' is imported via 'Exporter', which reexports definition from 'Exportee'}}
+
+public func useInDefaultValue(_ a: Int = funcUsedFromDefaultValue()) {}
+// expected-remark @-1 {{struct 'Int' is imported via 'Swift'}}
+// expected-remark @-2 {{global function 'funcUsedFromDefaultValue()' is imported via 'DepUsedFromInlinableCode'}}
+
+public func genericType(_ a: Array<TypeUsedInSignature>) {}
+// expected-remark @-1 {{generic struct 'Array' is imported via 'Swift'}}
+// expected-remark @-2 {{struct 'TypeUsedInSignature' is imported via 'DepUsedInSignature'}}
+
+public func protocolComposition(_ a: any ComposedProtoA & ComposedProtoB) {}
+// expected-remark @-1 {{protocol 'ComposedProtoA' is imported via 'DepUsedInSignature'}}
+// expected-remark @-2 {{protocol 'ComposedProtoB' is imported via 'DepUsedInSignature'}}
+
+public func useConformance(_ a: any Proto = ConformingType()) {}
+// expected-remark @-1 {{protocol 'Proto' is imported via 'ConformanceBaseTypes'}}
+// expected-remark @-2 {{conformance of 'ConformingType' to protocol 'Proto' used here is imported via 'ConformanceDefinition'}}
+// expected-remark @-3 {{struct 'ConformingType' is imported via 'ConformanceBaseTypes'}}
+// expected-remark @-4 {{initializer 'init()' is imported via 'ConformanceBaseTypes'}}
```

Comment:
Is this supposed to work with an extension member as well or will it be supported later? Would be nice to cover that in the test as well.

Reply:
I'll look into adding extensions to the tests. There are a few use cases missing here, I'm not so worried about testing wide as this feature currently piggy backs on the existing exportability checks. And testing narrow allows to check that each associated import was correctly marked as being used in API, ideally each import would be referenced only once. This test goes half-way, I expect I'll have to revisit this test and improve it.